### PR TITLE
Add BIP371 fields

### PR DESCRIPTION
**Pull request by brandonblack**

### CODE REVIEW

**Comment by brandonblack** [comment]
Related to https://github.com/bitcoinjs/bitcoinjs-lib/pull/1742

**Comment by brandonblack** [comment]
One of the vectors from BIP-0371 had a bug, so the vectors imported here include the fix for that bug: https://github.com/bitcoin/bips/pull/1296

**Comment by junderw** [comment]
A quick glance looks good.

Would also like @motorina0 to look over it.

**Comment by motorina0** [comment]
> Would also like @motorina0 to look over it.

I will check next Mon/Tue.

File: ts_src/lib/converter/input/tapKeySig.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { KeyValue, TapKeySig } from '../../interfaces';
+import { InputTypes } from '../../typeFields';
+
+export function decode(keyVal: KeyValue): TapKeySig {
+  if (keyVal.key[0] !== InputTypes.TAP_KEY_SIG) {
```
Comment (motorina0) [inline ts_src/lib/converter/input/tapKeySig.ts:5]: nit pick: check for  `keyVal.key.length ===  1`

**Comment by motorina0** [comment]
LGTM, however it needs more unit-tests.
 - coverage does not meet global threshold (90%)
 - github workflows should be configured (see `bitcoinjs-lib`)

File: ts_src/lib/converter/input/tapKeySig.ts (hunk 1 above)
Comment (brandonblack) [reply ts_src/lib/converter/input/tapKeySig.ts:5]: Fixed.

**Comment by brandonblack** [comment]
Alright. Tests added, GH workflow configured at least somewhat.

Thanks!

**Comment by junderw** [comment]
Also remove travis config.

File: ts_src/lib/converter/input/tapScriptSig.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    (data.signature.length === 64 || data.signature.length === 65)
+  );
+}
+
+export function canAddToArray(
+  array: TapScriptSig[],
+  item: TapScriptSig,
+  dupeSet: Set<string>,
+): boolean {
+  const dupeString = item.pubkey.toString('hex');
```
Comment (motorina0) [inline ts_src/lib/converter/input/tapScriptSig.ts:57]: - a pub key is allowed for more leafs
- shouldn't the leaf hash be used also to determine duplicates?

**Comment by motorina0** [comment]
- in `package.json` the `test` script should also run the coverage (` && npm run coverage`)

**Comment by motorina0** [comment]
- Business Logic side - all looks good 👍 

- Unit Test side, I'm trying to understand the  original vs new `twice` logic:
  - the original logic made two calls with a one item array, the second (twice) call altered the value for some fields
  - with the new logic
     - the old `twice` flag looses its meaning (second call has the same inputs, because data is not altered)
        - is this a regression in the tests?
     - the new tests make one call with a **two** item array (different values for each item), instead of two calls
        - is this intended?

**Comment by brandonblack** [comment]
The old twice logic was being (IMO) abused to do 2 things:
1. For valid cases, it was modifying specific keys to squeeze them through successfully
2. For invalid cases, it was modifying nothing to deliberately fail

Now twice is only used for invalid cases to ensure that duplicates don't sneak through, and for valid cases we explicitly put in the duplicates.

There is a slight difference between two calls with different values and 1 with 2 different values. That can be seen in `lib/utils.ts`. When it's one call, the `dupeSet` detects the duplicates, but when it's 2 calls, the array filter detects it.

Overall the new way allows for more flexibility (it's now possible by adding fixtures to fully exercise both paths).

**Review by motorina0 — APPROVED** [review]

**Pull request merged**